storage: Trim overlapping timestamp intervals in TimestampCache

nvb opened the pull request:
Fixes #4616 and closes #4700. This is an alternative to #4699.

Previously, we only removed timestamp intervals from the `TimestampCache`
during addition when a newer timestamp interval itself fully superseded
the earlier timestamp interval. This failed to take into account that only
the most recent timestamps for any given interval was needed to be
retained by the cache (we only need to keep the upper edge). With this new
assumption, newer timestamp intervals can be allowed to trim/split/or
delete any intervals that come before them and that they overlap. By
performing this reduction in size to any intervals that are overlapped,
we increase the chance that future timestamps will fully supersede past
timestamps, permitting their removal. It also seems to increase the
chance that past intervals can be replaced in-place with future intervals
by simply replacing their cache value. This allows us to more precisely
determine when a timestamp interval is still contributing to the cache's
maximum timestamp for some segment of the keyspace, and if it's not, remove
it on the spot instead of waiting for the `MinTSCacheWindow` (currently set
to 10 seconds) before evicting the interval.

This missing optimization during `TimestampCache` insertion meant that the
`TimestampCache` grew to extreme sizes, often accounting for over **300 MB**
per-node, which meant that it accounted for about **88%** of all memory
usage for Cockroach nodes. With this new optimization now being diligent
about removing timestamp intervals from the timestamp cache as soon as
they can be removed, we are able to reduce the upper bound for the per-node
`TimestampCache` size to about **3 MB**, a **100 x** reduction in size.
This means that we are able to reduce the total Cockroach node memory usage
from around **360 MB** to around **60 MB**!

Another key here is that when we adjust intervals in the `TimestampCache`,
we **NEED** to make sure the ordering in the FIFO queue stays consistent
with interval timestamps. For instance, if we don't move the interval back
in the queue in the value replacement case, intervals will get to a point
where their eviction is blocked on the newly replaced interval. This can
lead to memory growth of upwards of **40 MB**.

Because this optimization does more work during `TimestampCache` insertion, it
is expected that it will hurt performance slightly. That said, if benchmarks are
run as is, they will lead to misleading/incorrect results. This is because the
`MinTSCacheWindow` is currently set to 10 seconds, meaning that pre-optimization
benchmarks never had to perform costly timestamp cache deletions during eviction
because they always finished before this time. Keep in mind while looking at them
that after the 10 seconds, this `MinTSCacheWindow` eviction was previously still
happening synchronously during timestamp insertion, it was just pushing the cost
of deletion 10 seconds into the future. The only real discrepancies in
work comes from the need to now occasionally split intervals, which
results in an additional insertion and an additional deletion from the
cache.

Comparing this PR to #4699:

CPU performance
- Less computation needed on insert because fewer intervals will
  overlap since they will be trimmed as we go.
- This difference can be seen in each change's respective benchmarks.

Memory performance
- Average TSCache size is **3 MB** here vs. **4 MB** there
- #4699 will always have fewer TS cache entries because it does not chop
  up intervals that were superseded in the middle. These chops require
  inserting a whole new cache entry which requires allocations. However,
  only **.05%** of insertions are split in this PR, so it doesn't seem like this
  is causing an issue.
- **32%** of insertions resulting in a direct replacement here vs **~0%** there.
- **.25%** of insertions resulting in a deletion here vs **25%** there.

Perf Impact:

```
name                      old time/op    new time/op    delta
Bank_Cockroach-4             313µs ± 2%     339µs ±10%   +8.29%         (p=0.016 n=8+10)
Select1_Cockroach-4         52.7µs ± 1%    52.6µs ± 1%     ~            (p=0.549 n=9+10)
Select2_Cockroach-4          791µs ± 1%     775µs ± 1%   -2.07%          (p=0.000 n=9+9)
Insert1_Cockroach-4          393µs ±17%     371µs ± 2%     ~           (p=0.190 n=10+10)
Insert10_Cockroach-4         660µs ± 2%     650µs ± 0%   -1.48%          (p=0.001 n=8+9)
Insert100_Cockroach-4       3.39ms ± 1%    3.33ms ± 0%   -1.51%         (p=0.000 n=10+9)
Insert1000_Cockroach-4      33.1ms ± 7%    31.8ms ± 1%   -4.03%         (p=0.000 n=9+10)
Update1_Cockroach-4          591µs ± 9%     566µs ± 1%     ~             (p=0.063 n=9+9)
Update10_Cockroach-4        1.11ms ± 1%    1.10ms ± 0%     ~             (p=0.321 n=9+8)
Update100_Cockroach-4       6.12ms ± 5%    5.98ms ± 1%   -2.30%          (p=0.002 n=8+8)
Update1000_Cockroach-4      55.1ms ±10%    50.9ms ± 1%   -7.73%         (p=0.000 n=9+10)
Delete1_Cockroach-4          785µs ±22%     695µs ± 2%  -11.46%        (p=0.035 n=10+10)
Delete10_Cockroach-4        1.98ms ± 4%    2.09ms ± 3%   +5.36%        (p=0.000 n=10+10)
Delete100_Cockroach-4       18.0ms ±11%    17.3ms ± 1%     ~           (p=0.143 n=10+10)
Delete1000_Cockroach-4       144ms ± 5%     162ms ± 1%  +12.91%         (p=0.000 n=9+10)
Scan1_Cockroach-4            199µs ± 9%     183µs ± 1%   -7.83%        (p=0.015 n=10+10)
Scan10_Cockroach-4           241µs ±15%     215µs ± 1%  -10.74%         (p=0.000 n=10+9)
Scan100_Cockroach-4          478µs ± 1%     473µs ± 1%   -1.21%         (p=0.000 n=8+10)
Scan1000_Cockroach-4        3.12ms ± 5%    2.97ms ± 1%   -4.93%        (p=0.000 n=10+10)
Scan10000_Cockroach-4       32.2ms ± 3%    30.4ms ± 1%   -5.42%         (p=0.000 n=10+9)
PgbenchExec_Cockroach-4     2.83ms ±16%    2.75ms ± 5%     ~           (p=0.684 n=10+10)
PgbenchQuery_Cockroach-4    2.68ms ± 5%    2.63ms ± 3%     ~            (p=0.173 n=8+10)

name                      old alloc/op   new alloc/op   delta
Bank_Cockroach-4            69.1kB ± 1%    69.2kB ± 1%     ~           (p=0.403 n=10+10)
Select1_Cockroach-4         3.90kB ± 0%    3.90kB ± 0%     ~            (p=1.172 n=9+10)
Select2_Cockroach-4         80.5kB ± 0%    80.5kB ± 0%   -0.04%        (p=0.022 n=10+10)
Insert1_Cockroach-4         22.8kB ± 0%    22.7kB ± 0%   -0.12%        (p=0.023 n=10+10)
Insert10_Cockroach-4        66.8kB ± 0%    66.8kB ± 0%     ~           (p=0.565 n=10+10)
Insert100_Cockroach-4        506kB ± 0%     505kB ± 0%     ~            (p=0.156 n=10+9)
Insert1000_Cockroach-4      4.53MB ± 1%    4.51MB ± 0%   -0.51%         (p=0.008 n=10+9)
Update1_Cockroach-4         36.4kB ± 0%    36.4kB ± 0%     ~           (p=0.739 n=10+10)
Update10_Cockroach-4        98.8kB ± 0%    98.8kB ± 0%     ~           (p=0.382 n=10+10)
Update100_Cockroach-4        701kB ± 0%     699kB ± 0%   -0.30%        (p=0.023 n=10+10)
Update1000_Cockroach-4      6.05MB ± 0%    6.06MB ± 1%     ~            (p=0.780 n=9+10)
Delete1_Cockroach-4         35.3kB ± 0%    35.6kB ± 0%   +0.79%         (p=0.000 n=10+9)
Delete10_Cockroach-4         105kB ± 0%     108kB ± 0%   +2.56%         (p=0.000 n=10+9)
Delete100_Cockroach-4        788kB ± 0%     815kB ± 0%   +3.33%          (p=0.000 n=9+9)
Delete1000_Cockroach-4      7.13MB ± 1%    7.37MB ± 1%   +3.33%        (p=0.000 n=10+10)
Scan1_Cockroach-4           13.1kB ± 0%    13.1kB ± 0%     ~           (p=0.810 n=10+10)
Scan10_Cockroach-4          17.0kB ± 0%    17.0kB ± 0%   -0.15%         (p=0.041 n=9+10)
Scan100_Cockroach-4         49.2kB ± 0%    49.2kB ± 0%     ~           (p=0.955 n=10+10)
Scan1000_Cockroach-4         331kB ± 0%     331kB ± 0%   -0.04%        (p=0.000 n=10+10)
Scan10000_Cockroach-4       5.74MB ± 0%    5.72MB ± 0%   -0.37%        (p=0.000 n=10+10)
PgbenchExec_Cockroach-4      169kB ± 9%     174kB ±12%     ~             (p=0.387 n=9+9)
PgbenchQuery_Cockroach-4     213kB ± 0%     213kB ± 0%   +0.07%         (p=0.001 n=8+10)

name                      old allocs/op  new allocs/op  delta
Bank_Cockroach-4             1.67k ± 0%     1.68k ± 1%     ~            (p=0.633 n=8+10)
Select1_Cockroach-4           97.0 ± 0%      97.0 ± 0%     ~     (all samples are equal)
Select2_Cockroach-4          2.70k ± 0%     2.70k ± 0%     ~     (all samples are equal)
Insert1_Cockroach-4            352 ± 0%       352 ± 0%   -0.14%         (p=0.022 n=9+10)
Insert10_Cockroach-4           903 ± 0%       902 ± 0%     ~           (p=0.720 n=10+10)
Insert100_Cockroach-4        6.17k ± 0%     6.17k ± 0%     ~            (p=0.249 n=10+9)
Insert1000_Cockroach-4       58.8k ± 0%     58.7k ± 0%   -0.20%         (p=0.020 n=10+9)
Update1_Cockroach-4            712 ± 0%       712 ± 0%     ~     (all samples are equal)
Update10_Cockroach-4         1.40k ± 0%     1.40k ± 0%     ~           (p=0.720 n=10+10)
Update100_Cockroach-4        7.93k ± 0%     7.90k ± 0%     ~           (p=0.057 n=10+10)
Update1000_Cockroach-4       70.0k ± 0%     70.0k ± 1%     ~            (p=1.000 n=9+10)
Delete1_Cockroach-4            681 ± 0%       684 ± 0%   +0.46%        (p=0.000 n=10+10)
Delete10_Cockroach-4         1.66k ± 0%     1.69k ± 0%   +1.79%        (p=0.000 n=10+10)
Delete100_Cockroach-4        11.1k ± 0%     11.4k ± 0%   +2.66%          (p=0.000 n=8+8)
Delete1000_Cockroach-4        106k ± 0%      108k ± 0%   +2.52%        (p=0.000 n=10+10)
Scan1_Cockroach-4              272 ± 0%       272 ± 0%     ~     (all samples are equal)
Scan10_Cockroach-4             352 ± 0%       352 ± 0%     ~           (p=0.577 n=10+10)
Scan100_Cockroach-4          1.08k ± 0%     1.08k ± 0%     ~           (p=1.000 n=10+10)
Scan1000_Cockroach-4         8.29k ± 0%     8.29k ± 0%     ~     (all samples are equal)
Scan10000_Cockroach-4        80.4k ± 0%     80.4k ± 0%   +0.01%         (p=0.000 n=10+9)
PgbenchExec_Cockroach-4      3.29k ±21%     3.37k ±17%     ~           (p=0.436 n=10+10)
PgbenchQuery_Cockroach-4     3.87k ± 0%     3.88k ± 0%   +0.23%         (p=0.000 n=8+10)
```

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.svg" height="40" alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/4789)

<!-- Reviewable:end -->

nvb committed:
storage: Trim overlapping timestamp intervals in TimestampCache

Fixes #4616 and closes #4700. This is an alternative to #4699.

Previously, we only removed timestamp intervals from the `TimestampCache`
during addition when a newer timestamp interval itself fully superseded
the earlier timestamp interval. This failed to take into account that only
the most recent timestamps for any given interval was needed to be
retained by the cache (we only need to keep the upper edge). With this new
assumption, newer timestamp intervals can be allowed to trim/split/or
delete any intervals that come before them and that they overlap. By
performing this reduction in size to any intervals that are overlapped,
we increase the chance that future timestamps will fully supersede past
timestamps, permitting their removal. It also seems to increase the
chance that past intervals can be replaced in-place with future intervals
by simply replacing their cache value. This allows us to more precisely
determine when a timestamp interval is still contributing to the cache's
maximum timestamp for some segment of the keyspace, and if it's not, remove
it on the spot instead of waiting for the `MinTSCacheWindow` (currently set
to 10 seconds) before evicting the interval.

This missing optimization during `TimestampCache` insertion meant that the
`TimestampCache` grew to extreme sizes, often accounting for over **300 MB**
per-node, which meant that it accounted for about **88%** of all memory
usage for Cockroach nodes. With this new optimization now being diligent
about removing timestamp intervals from the timestamp cache as soon as
they can be removed, we are able to reduce the upper bound for the per-node
`TimestampCache` size to about **3 MB**, a **100 x** reduction in size.
This means that we are able to reduce the total Cockroach node memory usage
from around **360 MB** to around **60 MB**!

Another key here is that when we adjust intervals in the `TimestampCache`,
we **NEED** to make sure the ordering in the FIFO queue stays consistent
with interval timestamps. For instance, if we don't move the interval back
in the queue in the value replacement case, intervals will get to a point
where their eviction is blocked on the newly replaced interval. This can
lead to memory growth of upwards of 40 MB!

Because this optimization does more work during `TimestampCache` insertion, it
is expected that it will hurt performance slightly. That said, if benchmarks are
run as is, they will lead to misleading/incorrect results. This is because the
`MinTSCacheWindow` is currently set to 10 seconds, meaning that pre-optimization
benchmarks never had to perform costly timestamp cache deletions during eviction
because they always finished before this time. Keep in mind while looking at them
that after the 10 seconds, this `MinTSCacheWindow` eviction was previously still
happening synchronously during timestamp insertion, it was just pushing the cost
of deletion 10 seconds into the future. The only real discrepancies in
work comes from the need to now occasionally split intervals, which
results in an additional insertion and an additional deletion from the
cache.

Comparing this PR to #4699:

CPU performance
- Less computation needed on insert because fewer intervals will
  overlap since they will be trimmed as we go.
- This difference can be seen in each change's respective benchmarks.

Memory performance
- Average TSCache size is **3 MB** here vs. **4 MB** there
- #4699 will always have fewer TS cache entries because it does not chop
  up intervals that were superseded in the middle. These chops require
  inserting a whole new cache entry which requires allocations. However,
  only **.05%** of insertions are split in this PR, so it doesn't seem like this
  is causing an issue.
- **32%** of insertions resulting in a direct replacement here vs **~0%** there.
- **.25%** of insertions resulting in a deletion here vs **25%** there.

Perf Impact:

```
name                      old time/op    new time/op    delta
Bank_Cockroach-4             313µs ± 2%     339µs ±10%   +8.29%         (p=0.016 n=8+10)
Select1_Cockroach-4         52.7µs ± 1%    52.6µs ± 1%     ~            (p=0.549 n=9+10)
Select2_Cockroach-4          791µs ± 1%     775µs ± 1%   -2.07%          (p=0.000 n=9+9)
Insert1_Cockroach-4          393µs ±17%     371µs ± 2%     ~           (p=0.190 n=10+10)
Insert10_Cockroach-4         660µs ± 2%     650µs ± 0%   -1.48%          (p=0.001 n=8+9)
Insert100_Cockroach-4       3.39ms ± 1%    3.33ms ± 0%   -1.51%         (p=0.000 n=10+9)
Insert1000_Cockroach-4      33.1ms ± 7%    31.8ms ± 1%   -4.03%         (p=0.000 n=9+10)
Update1_Cockroach-4          591µs ± 9%     566µs ± 1%     ~             (p=0.063 n=9+9)
Update10_Cockroach-4        1.11ms ± 1%    1.10ms ± 0%     ~             (p=0.321 n=9+8)
Update100_Cockroach-4       6.12ms ± 5%    5.98ms ± 1%   -2.30%          (p=0.002 n=8+8)
Update1000_Cockroach-4      55.1ms ±10%    50.9ms ± 1%   -7.73%         (p=0.000 n=9+10)
Delete1_Cockroach-4          785µs ±22%     695µs ± 2%  -11.46%        (p=0.035 n=10+10)
Delete10_Cockroach-4        1.98ms ± 4%    2.09ms ± 3%   +5.36%        (p=0.000 n=10+10)
Delete100_Cockroach-4       18.0ms ±11%    17.3ms ± 1%     ~           (p=0.143 n=10+10)
Delete1000_Cockroach-4       144ms ± 5%     162ms ± 1%  +12.91%         (p=0.000 n=9+10)
Scan1_Cockroach-4            199µs ± 9%     183µs ± 1%   -7.83%        (p=0.015 n=10+10)
Scan10_Cockroach-4           241µs ±15%     215µs ± 1%  -10.74%         (p=0.000 n=10+9)
Scan100_Cockroach-4          478µs ± 1%     473µs ± 1%   -1.21%         (p=0.000 n=8+10)
Scan1000_Cockroach-4        3.12ms ± 5%    2.97ms ± 1%   -4.93%        (p=0.000 n=10+10)
Scan10000_Cockroach-4       32.2ms ± 3%    30.4ms ± 1%   -5.42%         (p=0.000 n=10+9)
PgbenchExec_Cockroach-4     2.83ms ±16%    2.75ms ± 5%     ~           (p=0.684 n=10+10)
PgbenchQuery_Cockroach-4    2.68ms ± 5%    2.63ms ± 3%     ~            (p=0.173 n=8+10)

name                      old alloc/op   new alloc/op   delta
Bank_Cockroach-4            69.1kB ± 1%    69.2kB ± 1%     ~           (p=0.403 n=10+10)
Select1_Cockroach-4         3.90kB ± 0%    3.90kB ± 0%     ~            (p=1.172 n=9+10)
Select2_Cockroach-4         80.5kB ± 0%    80.5kB ± 0%   -0.04%        (p=0.022 n=10+10)
Insert1_Cockroach-4         22.8kB ± 0%    22.7kB ± 0%   -0.12%        (p=0.023 n=10+10)
Insert10_Cockroach-4        66.8kB ± 0%    66.8kB ± 0%     ~           (p=0.565 n=10+10)
Insert100_Cockroach-4        506kB ± 0%     505kB ± 0%     ~            (p=0.156 n=10+9)
Insert1000_Cockroach-4      4.53MB ± 1%    4.51MB ± 0%   -0.51%         (p=0.008 n=10+9)
Update1_Cockroach-4         36.4kB ± 0%    36.4kB ± 0%     ~           (p=0.739 n=10+10)
Update10_Cockroach-4        98.8kB ± 0%    98.8kB ± 0%     ~           (p=0.382 n=10+10)
Update100_Cockroach-4        701kB ± 0%     699kB ± 0%   -0.30%        (p=0.023 n=10+10)
Update1000_Cockroach-4      6.05MB ± 0%    6.06MB ± 1%     ~            (p=0.780 n=9+10)
Delete1_Cockroach-4         35.3kB ± 0%    35.6kB ± 0%   +0.79%         (p=0.000 n=10+9)
Delete10_Cockroach-4         105kB ± 0%     108kB ± 0%   +2.56%         (p=0.000 n=10+9)
Delete100_Cockroach-4        788kB ± 0%     815kB ± 0%   +3.33%          (p=0.000 n=9+9)
Delete1000_Cockroach-4      7.13MB ± 1%    7.37MB ± 1%   +3.33%        (p=0.000 n=10+10)
Scan1_Cockroach-4           13.1kB ± 0%    13.1kB ± 0%     ~           (p=0.810 n=10+10)
Scan10_Cockroach-4          17.0kB ± 0%    17.0kB ± 0%   -0.15%         (p=0.041 n=9+10)
Scan100_Cockroach-4         49.2kB ± 0%    49.2kB ± 0%     ~           (p=0.955 n=10+10)
Scan1000_Cockroach-4         331kB ± 0%     331kB ± 0%   -0.04%        (p=0.000 n=10+10)
Scan10000_Cockroach-4       5.74MB ± 0%    5.72MB ± 0%   -0.37%        (p=0.000 n=10+10)
PgbenchExec_Cockroach-4      169kB ± 9%     174kB ±12%     ~             (p=0.387 n=9+9)
PgbenchQuery_Cockroach-4     213kB ± 0%     213kB ± 0%   +0.07%         (p=0.001 n=8+10)

name                      old allocs/op  new allocs/op  delta
Bank_Cockroach-4             1.67k ± 0%     1.68k ± 1%     ~            (p=0.633 n=8+10)
Select1_Cockroach-4           97.0 ± 0%      97.0 ± 0%     ~     (all samples are equal)
Select2_Cockroach-4          2.70k ± 0%     2.70k ± 0%     ~     (all samples are equal)
Insert1_Cockroach-4            352 ± 0%       352 ± 0%   -0.14%         (p=0.022 n=9+10)
Insert10_Cockroach-4           903 ± 0%       902 ± 0%     ~           (p=0.720 n=10+10)
Insert100_Cockroach-4        6.17k ± 0%     6.17k ± 0%     ~            (p=0.249 n=10+9)
Insert1000_Cockroach-4       58.8k ± 0%     58.7k ± 0%   -0.20%         (p=0.020 n=10+9)
Update1_Cockroach-4            712 ± 0%       712 ± 0%     ~     (all samples are equal)
Update10_Cockroach-4         1.40k ± 0%     1.40k ± 0%     ~           (p=0.720 n=10+10)
Update100_Cockroach-4        7.93k ± 0%     7.90k ± 0%     ~           (p=0.057 n=10+10)
Update1000_Cockroach-4       70.0k ± 0%     70.0k ± 1%     ~            (p=1.000 n=9+10)
Delete1_Cockroach-4            681 ± 0%       684 ± 0%   +0.46%        (p=0.000 n=10+10)
Delete10_Cockroach-4         1.66k ± 0%     1.69k ± 0%   +1.79%        (p=0.000 n=10+10)
Delete100_Cockroach-4        11.1k ± 0%     11.4k ± 0%   +2.66%          (p=0.000 n=8+8)
Delete1000_Cockroach-4        106k ± 0%      108k ± 0%   +2.52%        (p=0.000 n=10+10)
Scan1_Cockroach-4              272 ± 0%       272 ± 0%     ~     (all samples are equal)
Scan10_Cockroach-4             352 ± 0%       352 ± 0%     ~           (p=0.577 n=10+10)
Scan100_Cockroach-4          1.08k ± 0%     1.08k ± 0%     ~           (p=1.000 n=10+10)
Scan1000_Cockroach-4         8.29k ± 0%     8.29k ± 0%     ~     (all samples are equal)
Scan10000_Cockroach-4        80.4k ± 0%     80.4k ± 0%   +0.01%         (p=0.000 n=10+9)
PgbenchExec_Cockroach-4      3.29k ±21%     3.37k ±17%     ~           (p=0.436 n=10+10)
PgbenchQuery_Cockroach-4     3.87k ± 0%     3.88k ± 0%   +0.23%         (p=0.000 n=8+10)
```

nvb committed:
storage: IntervalCache keeps ordering correct when adjusting keys

When we adjust intervals in the `TimestampCache`, we need to make sure
the ordering in the FIFO queue stays consistent with interval timestamps.
For instace, if we don't move the interval back in the queue in the value
replacement case, intervals will get to a point where their eviction
is blocked on the newly replaced interval. This can lead to memory
growth of upwards of **40 MB**!

This change includes the following `util/cache` methods to help with
this:
- (*baseCache).AddEntryAfter
- (*baseCache).MoveToEnd

nvb committed:
util/cache: Allow cache DelEntry to avoid lookup

This change makes a few changes to `util/cache.baseCache` to allow
direct usageof `*cache.Entry` values when possible when deleting entries.
This allows us to avoid Entry lookup.

petermattis commented on storage/timestamp_cache.go:
```diff
@@ -114,61 +120,151 @@ func (tc *TimestampCache) Add(start, end roachpb.Key, timestamp roachpb.Timestam
-				tc.cache.Del(o.Key) // delete existing key; this cache entry supersedes.
+			} else {
+				// The existing interval has a timestamp greater than the new interval.
+				if rg == nil {
```
I don't think the range group is necessary here. Similar to the above case of slicing the old interval, the new interval is being sliced. If the new interval is sliced into two pieces, you can always add the first piece to the cache as it is guaranteed to be before the existing overlap we're looking at. I doubt this will affect performance at all, but will make this else-branch more symmetric with the if-branch.

nvanbenschoten replied:
Ah yeah, I forgot the ordering of the overlapping intervals from `GetOverlaps` was deterministic and ordered by start key. Just made the change using that assumption and it makes things a good bit cleaner.

petermattis commented on storage/timestamp_cache.go:
```diff
@@ -114,61 +120,151 @@ func (tc *TimestampCache) Add(start, end roachpb.Key, timestamp roachpb.Timestam
+					// New:     ----
+					// Old: ------------
+					//
+					// Res: ----    ----
```
Love these diagrams.

petermattis commented:
This looks better than #4699. Are the `TimestampCache` tests exercising all of the code paths in `TimestampCache.Add`? Lots of permutations to test there.

nvb committed:
util/cache: Fix baseCahce.Clear by walking linked list and dumping store

Before this change, there was a subtle bug where `IntervalCache.clear`
would fail to clear all entries from the cache. This issue came about
when interval keys had been modified in the cache in certain ways, such
that when deleting entries from the underlying red-black tree (in an
in-order depth-first ordering provided by `tree.Do`), rotations occurred.
If a rotation occurred, the traversal could be messed up and nodes could
be missed.

This change fixes this issue by simply replacing the entire tree on
clear, which should also be more efficient. It also includes a repro test
case for the issue it fixes.

It follows this pattern with the `OrderedCache` and `UnorderedCache` by
adding a new `init` method to the `cacheStore` interface and removing
`del` from the interface. A mildly noteworthy part of this change is
that it requires us to now use the `baseCache`'s linked list even if
the eviction policy is set to `CacheNone`.

nvanbenschoten commented:
Just added some test cases that exercise all code paths in `TimestampCache.Add` (confirmed with code coverage). They also verify that the resulting `TimestampCache`'s max intervals are independent of insertion order and only dependent on insertion timestamp.

Creating these tests revealed a really subtle bug with `IntervalCache.Clear` having to do with individual deletions rotating parts of the RB-tree and messing up our depth-first traversal, which I fixed in a separate commit. Let me know what you think of the issue and the fix.

petermattis commented:
LGTM

---

Review status: 0 of 8 files reviewed at latest revision, 3 unresolved discussions.

---

*[storage/timestamp_cache.go, line 126 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4789#-KBwztIVu-v03Fp7Qlg3:-KBwztIVu-v03Fp7Qlg4:594227489) ([raw file](https://github.com/cockroachdb/cockroach/blob/2cac2af0a0f20fd866e8510952b832995a7b4938/storage/timestamp_cache.go#L126)):*
I believe using a closure likes this requires an allocation to capture `{cache,timestamp,txnID}`. It would be mildly better to create an `addRange()` method that took those variables as parameters.

---

*[util/cache/cache.go, line 483 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4789#-KBx0zinw8VFKokNz6-n:-KBx0zio4WIHvKlkvOPm:1097507522) ([raw file](https://github.com/cockroachdb/cockroach/blob/2cac2af0a0f20fd866e8510952b832995a7b4938/util/cache/cache.go#L483)):*
Rather than walking the tree like this, perhaps `baseCache.Clear()` should walk the linked list and call `OnEvicted` before resetting the linked-list. Then we could avoid the traversal altogether when `OnEvicted == nil` (the common case).

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4789#-:-KBx0ryCcjLR-j6zuqp_:676834233)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
Review status: 0 of 8 files reviewed at latest revision, 3 unresolved discussions.

---

*[storage/timestamp_cache.go, line 126 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4789#-KBwztIVu-v03Fp7Qlg3:-KBx3x7pWsc_C_weV6TU:635138701) ([raw file](https://github.com/cockroachdb/cockroach/blob/2cac2af0a0f20fd866e8510952b832995a7b4938/storage/timestamp_cache.go#L126)):*
So I looked into if that was the case when I wrote it, and I *believe* escape analysis will be able to tell that the closure doesn't escape and avoid the allocation. I stumbled upon https://github.com/golang/go/commit/9fe60801aeca801010b42e1dd7ad57a173dc9740, which I think would apply here. I also looked at the escape analysis `gcflags` and it didn't mention anything about those three variables being moved to the heap.

---

*[util/cache/cache.go, line 483 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4789#-KBx0zinw8VFKokNz6-n:-KBx6wdi-4IQ74wZYjrJ:1692988062) ([raw file](https://github.com/cockroachdb/cockroach/blob/2cac2af0a0f20fd866e8510952b832995a7b4938/util/cache/cache.go#L483)):*
How's that? I don't see any compelling reason not to just dump the entire list and tree on clear. I called the method to reset the cache impls `init` to remain consistent with the linked list.

On a side note, a mildly noteworthy part of that change was that it requires us to now use the linked list even if the eviction policy is none. It doesn't seem like a huge deal, but it's something to consider. 

I'll squash this in with the previous commit if you agree with the approach.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4789)*
<!-- Sent from Reviewable.io -->

petermattis commented:
LGTM

---

Review status: 0 of 8 files reviewed at latest revision, 2 unresolved discussions.

---

*[util/cache/cache.go, line 483 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4789#-KBx0zinw8VFKokNz6-n:-KBxDJn58TVLCHnyw_Ud:1637927125) ([raw file](https://github.com/cockroachdb/cockroach/blob/2cac2af0a0f20fd866e8510952b832995a7b4938/util/cache/cache.go#L483)):*
Seems fine to me. 

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4789#-:-KBxDOxuDELEgecxGSpJ:676834233)*
<!-- Sent from Reviewable.io -->
